
message OffsetTable {
    required int64 start_addr = 1;
    repeated int64 table_offsets = 2;
    repeated int64 destinations = 3;
}

message JumpTbl {
    repeated    int64       table_entries = 1;
    required    int32       zero_offset = 2;
    optional    int64       offset_from_data = 3;
}

message JumpIndexTbl {
    required    bytes       table_entries = 1;
    required    int32       zero_offset = 2;
}

message Reference {
    enum TargetType {
        CodeTarget = 0;
        DataTarget = 1;
    }

    enum OperandType {
        ImmediateOperand = 0;
        MemoryOperand = 1;  // Absolute and PC-relative accesses.
        MemoryDisplacementOperand = 2;
        ControlFlowOperand = 3;  // E.g. Function call target.
    }

    enum Location {
        Internal = 0;
        External = 1;
    }

    required    TargetType  target_type = 1;
    required    OperandType operand_type = 2;
    required    Location    location = 3;
    required    int64       address = 4;
    optional    string      name = 5;
}

message Instruction {
    required    bytes             inst_bytes = 1;
    required    int64             inst_addr = 2;

    repeated    Reference         refs = 3;

    optional    JumpTbl           jump_table = 4;
    optional    JumpIndexTbl      jump_index_table = 5;
    optional    bool              local_noreturn = 6;
    optional    int64             offset_table_addr = 7;
}

message Block {
    repeated    Instruction insts = 1;
    required    int64       base_address = 2;
    //the block_follows list is used to rebuild the basic block CFG, so it 
    //should contain the list of base_addresses for blocks following this 
    //block in the CFG
    repeated    int64       block_follows = 3;
}

message Function {
    repeated    Block   blocks = 1;
    required    int64   entry_address = 2;
    optional    string  symbol_name = 3;
}

message ExternalFunction {
    enum CallingConvention {
      CallerCleanup = 0;
      CalleeCleanup = 1;
      FastCall      = 2;
      McsemaCall    = 3;
    }

    required    string            symbol_name = 1;
    required    CallingConvention calling_convention = 2;
    required    bool              has_return = 3;
    required    bool              no_return = 4;
    required    int32             argument_count = 5;
    required    bool              is_weak = 6;
    optional    string            signature = 7;
}

message ExternalData {
    required    string      symbol_name = 1;
    required    int32       data_size = 2;
    required    bool        is_weak = 3;
}


message DataSymbol {
    required int64 base_address = 1;
    required string symbol_name = 2;
    required int32 symbol_size = 3;
}

message Data {
    required    int64           base_address = 1;
    required    bytes           data = 2;
    repeated    DataSymbol      symbols = 3;    
    required    bool            read_only = 4;
}

message EntrySymbolExtra {
    required    int32                               entry_argc = 1;
    required    ExternalFunction.CallingConvention  entry_cconv = 2;
    required    bool                                does_return = 3;
}

message EntrySymbol {
    required    string                              entry_name = 1;
    required    int64                               entry_address = 2;
    optional    EntrySymbolExtra                    entry_extra = 3;
}

message Module {
    repeated    Function            internal_funcs = 1;
    repeated    ExternalFunction    external_funcs = 2;
    repeated    Data                internal_data = 3;
    required    string              module_name = 4;
    repeated    EntrySymbol         entries = 5;
    repeated    ExternalData        external_data = 6;
}
